### PR TITLE
Adding an is_abtract check in pauli_decompose so it does not break under jit (either jax.jit or catalyst.qjit)

### LINKED ISSUE #5817: [BUG] `qml.pauli_decompose` fails when being jit-ted
### Expected behavior

Pauli decompose works well outside of `jax.jit`

```python
dev = qml.device("lightning.qubit", wires=1)

#@jax.jit
def f(mat):
	coeffs, unitaries = qml.pauli_decompose(mat, check_hermitian=False).terms()
	return coeffs, unitaries

x = jnp.array([[1,2],[3,4] ])
print(f(x))

>>>
(Array([ 2.5+0.j ,  2.5+0.j , -0. -0.5j, -1.5+0.j ], dtype=complex128), [I(0), X(0), Y(0), Z(0)])
```

### Actual behavior

but fails when jit-ted:

```python
dev = qml.device("lightning.qubit", wires=1)

@jax.jit
def f(mat):
	coeffs, unitaries = qml.pauli_decompose(mat, check_hermitian=False).terms()
	return coeffs, unitaries

x = jnp.array([[1,2],[3,4] ])
print(f(x))

>>>
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/paul.wang/small_playgrounds_dump/qsvt_demo/pauli_decompose.py", line 20, in <module>
    print(f(x))
  File "/home/paul.wang/small_playgrounds_dump/qsvt_demo/pauli_decompose.py", line 16, in f
    coeffs, unitaries = qml.pauli_decompose(mat, check_hermitian=False).terms()
  File "/home/paul.wang/.local/lib/python3.10/site-packages/pennylane/pauli/conversion.py", line 323, in pauli_decompose
    coeffs, obs = _generalized_pauli_decompose(
  File "/home/paul.wang/.local/lib/python3.10/site-packages/pennylane/pauli/conversion.py", line 209, in _generalized_pauli_decompose
    if not qml.math.allclose(coefficient, 0):
jax.errors.TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
The error occurred while tracing the function f at /home/paul.wang/small_playgrounds_dump/qsvt_demo/pauli_decompose.py:14 for jit. This concrete value was not available in Python because it depends on the value of the argument mat.
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError
```

Note that the input matrix cannot be set as `static_argnum` as it is a (j)numpy array, which is not hashable. 

This error is the usual "you are using an abstract traced value as a concrete value" error in jax, which likely is because there is a conditional depending on the concrete value of the input matrix; however, turning off `check_hermitian` did not help. 

### Additional information

_No response_

### Source code

_No response_

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.37.0.dev0
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/paul.wang/.local/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: amazon-braket-pennylane-plugin, PennyLane-Catalyst, PennyLane-Lightning-Kokkos, PennyLane_Lightning

Platform info:           Linux-6.5.0-35-generic-x86_64-with-glibc2.35
Python version:          3.10.12
Numpy version:           1.26.4
Scipy version:           1.12.0
Installed devices:
- default.clifford (PennyLane-0.37.0.dev0)
- default.gaussian (PennyLane-0.37.0.dev0)
- default.mixed (PennyLane-0.37.0.dev0)
- default.qubit (PennyLane-0.37.0.dev0)
- default.qubit.autograd (PennyLane-0.37.0.dev0)
- default.qubit.jax (PennyLane-0.37.0.dev0)
- default.qubit.legacy (PennyLane-0.37.0.dev0)
- default.qubit.tf (PennyLane-0.37.0.dev0)
- default.qubit.torch (PennyLane-0.37.0.dev0)
- default.qutrit (PennyLane-0.37.0.dev0)
- default.qutrit.mixed (PennyLane-0.37.0.dev0)
- default.tensor (PennyLane-0.37.0.dev0)
- null.qubit (PennyLane-0.37.0.dev0)
- nvidia.custatevec (PennyLane-Catalyst-0.7.0.dev0)
- nvidia.cutensornet (PennyLane-Catalyst-0.7.0.dev0)
- oqc.cloud (PennyLane-Catalyst-0.7.0.dev0)
- softwareq.qpp (PennyLane-Catalyst-0.7.0.dev0)
- lightning.qubit (PennyLane_Lightning-0.36.0)
- braket.aws.ahs (amazon-braket-pennylane-plugin-1.27.0)
- braket.aws.qubit (amazon-braket-pennylane-plugin-1.27.0)
- braket.local.ahs (amazon-braket-pennylane-plugin-1.27.0)
- braket.local.qubit (amazon-braket-pennylane-plugin-1.27.0)
- lightning.kokkos (PennyLane-Lightning-Kokkos-0.32.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by paul0403**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** `pauli_decompose` currently fails under a `jax.jit` and `catalyst.qjit`

**Description of the Change:** An `is_abstract` check is added to avoid jit trying to compare an `isallclose` during compile time on an abstract tracer value

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:** closes #5817 

[sc-65302]
[sc-65344]

### CODE REVIEW

**Review by obliviateandsurrender — COMMENTED** [review]
Please add a `jax-jit` interface unittest for the changes as well!

**Comment by paul0403** [comment]
> A side effect of this change:
> 
> ```
> >>> x = jnp.array(np.diag([0, 0, 1]))
> >>> coeffs, unitaries = qml.pauli_decompose(x, check_hermitian=False).terms()
> >>> print(coeffs, unitaries)
> [ 0.25+0.j  0.25+0.j -0.25+0.j -0.25+0.j] [I(0) @ I(1), I(0) @ Z(1), Z(0) @ I(1), Z(0) @ Z(1)]
> ```
> 
> and after jitting,
> 
> ```
> >>> coeffs, unitaries = jax.jit(functools.partial(qml.pauli_decompose, check_hermitian=False))(x).terms()
> >>> print(coeffs, unitaries)
> [ 0.25+0.j  0.  +0.j  0.  +0.j  0.25+0.j  0.  +0.j  0.  +0.j  0.  +0.j
>   0.  +0.j  0.  +0.j  0.  +0.j -0.  +0.j  0.  +0.j -0.25+0.j  0.  +0.j
>   0.  +0.j -0.25+0.j] [I(0) @ I(1), I(0) @ X(1), I(0) @ Y(1), I(0) @ Z(1), X(0) @ I(1), X(0) @ X(1), X(0) @ Y(1), X(0) @ Z(1), Y(0) @ I(1), Y(0) @ X(1), Y(0) @ Y(1), Y(0) @ Z(1), Z(0) @ I(1), Z(0) @ X(1), Z(0) @ Y(1), Z(0) @ Z(1)]
> ```
> 
> but I guess since jit has to work with statically shaped arrays this is unavoidable.

Yes, jit would need to know the return shape in advance (especially in catalyst where everything will be lowered to mlir, which is strongly typed). There isn't any real workaround for this.

**Comment by astralcai** [comment]
> This change does not happen outside jitting right @astralcai ? i.e. pure core PL behavior stays the same? Because if it only occurs for jitting then I think we could safely assume any user of jit would be familiar with such jit-gotcha business

Yes, this wouldn't affect pure PL behaviour.

**Comment by obliviateandsurrender** [comment]
@astralcai would users be able to use `qml.simplify` to get rid of the zero components? For some reason, I couldn't 🤔

File: tests/pauli/test_conversion.py [hunk 1]
```diff
@@ -209,6 +209,18 @@ def test_wire_error(self, pauli):
         ):
             qml.pauli_decompose(hamiltonian, pauli=pauli, wire_order=wire_order)
 
+    def test_jit(self):
```
Comment (obliviateandsurrender) [inline tests/pauli/test_conversion.py:212]: Will need to add a `jax` marker for the test.

**Comment by astralcai** [comment]
> @astralcai would users be able to use `qml.simplify` to get rid of the zero components in that case? For some reason, I couldn't 🤔

`qml.simplify` does not remove terms with 0 coefficients.

**Review by obliviateandsurrender — APPROVED** [review]

**Comment by obliviateandsurrender** [comment]
Don't forget to add a `changelog` entry!

File: tests/pauli/test_conversion.py [hunk 2]
```diff
@@ -209,6 +209,19 @@ def test_wire_error(self, pauli):
...
             qml.pauli_decompose(hamiltonian, pauli=pauli, wire_order=wire_order)
 
+    @pytest.mark.jax
+    def test_jit(self):
+        """Test that pauli_decompose can be used inside a jit context"""
+        import jax
+
+        @jax.jit
+        def f(m):
+            coeffs, unitaries = qml.pauli_decompose(m, check_hermitian=False).terms()
```
Comment (astralcai) [inline tests/pauli/test_conversion.py:219]: What if the user did not include `check_hermitian=False`? Should we add the `is_abstract` check alongside `if check_hermitian` in `pauli_decompose`?

Comment (paul0403) [reply tests/pauli/test_conversion.py:219]: `check_hermitian=False` is definitely needed in this version. Let me try adding `is_abstract` along the hermiticity check as well

Comment (paul0403) [reply tests/pauli/test_conversion.py:219]: very good suggestion, I was actually running into the same abstract jax array difficulties with the hermiticity check in catalyst. This is definitely the place to fix that as well :100:

Comment (astralcai) [reply tests/pauli/test_conversion.py:219]: Now `check_hermitian=False` wouldn't be needed here I think?

Comment (paul0403) [reply tests/pauli/test_conversion.py:219]: It is still needed, as this input matrix is purposefully not hermitian (so turning it off would real-lize the coefficients and would be giving the wrong answers). I added another test for hermitian matrices, where `check_hermitian` kwarg is omitted

**Review by astralcai — APPROVED** [review]
LGTM

**Pull request merged**